### ISSUE TITLE
Event on_community_store_order: Variable not Updating, ORM Usage Bug

### ISSUE THREAD

**Issue by smansfield** [concrete5-community-store/community_store#85]
Note that all line numbers are for the file `src/CommunityStore/Order/Order.php` in it's current state on Github.

## Bug Description / How I Encountered this Bug
Recently, I have been trying to hook into the event `on_community_store_order` (dispatched on line 683) to run an action whenever a new user is added to concrete5 by community store. To run the action I need the user ID, contained in the cID value for the event object. Here is an example of how this would be used:
```PHP
Events::addListener('on_community_store_order', $event) {
    $cID = $event->getOrder()->getOrderID();
    ....
    // use cID, getUserByID($cID), etc.
}
```
I need to hook into this event instead of `on_user_add` because I only want this action to run when a user is added by community store, by ordering an item, and not when a user is added by anything else. The event object passed (OrderEvent from `src/CommunityStore/Order/OrderEvent.php`) contains all information about the order placed (available with the getOrder() method). However, there is a bug where the cID (customer ID) will be passed with a placeholder 0 if a new user has to be created for the order.

#### add() method (line 396)
The order begins processing in the add() method. The method does the following:

1. Populates a new Order object, named $order, using arguments passed to add() (line 433)

1. The user ID is set to 0, for at this point community store has not yet created a concrete5 user for the new order (line 434, also see getUserID() method of the Customer class)

1. $order is saved to the ORM (`$order->save()` on line 449)

1. The completeOrder() method is called. (line 481)

#### completeOrder() method (line 517)
This method does the following:

1. Creates a new concrete5 user for the order (line 565 - 616)

1. Associates the order object with the concrete5 user by calling the associateUser() method (line 639)

1. Dispatches the `on_community_store_order` with an event wrapper for the current order object (line 683)

At this point, when the event is dispatched, the cID value is still set to 0, even though the passed object was associated with the newly created concrete5 user in the associateUser() method! So here's the bug.

#### associateUser() method (line 923)
```PHP
public function associateUser($uID)
{
    $db = \Database::connection();
    $rows = $db->query("Update CommunityStoreOrders set cID=? where oID = ?", array($uID, $this->oID));
    return $rows;
}
```
All this method does is query the SQL database and update the cID column to the passed value, which is the user ID for the newly created concrete5 user. It does not save the cID to the order object, and it does not save the cID through the ORM. This is why when the event is dispatched, the cID in the SQL db is updated but the cID in the object is still 0. This is how the bug occurs.

## How is this a Bug?
- It prevents the `on_community_store_order` event from having consistent behavior because once the order object is recreated a second time (when the customer places a second order), it will properly return the cID as it is loaded from the database.
- It causes the event to be useless to me because I cannot use the cID on the first order of a new customer
- Typical fixes, such as loading the order object again after the event has been called, do not work. For example:
```PHP
use Concrete\Package\CommunityStore\Src\CommunityStore\Order\Order as Order;
Events::addListener('on_community_store_order', function($event) {
    $order = $event->getOrder();    // Get the order object from the event
    $oID = $order->getOrderID();    // Get the order ID from the order object
    $order = Order::getByID($oID);  // Load the order object again to refresh the values
    $cID = $order->getCustomerID(); // cID is still 0
    /** 
    * Does not work, $cID is still 0 even after re-populating the order object.
    * This is despite the fact that the cID has already updated in the SQL db.
    * I'm guessing this is because the order must be cached by the ORM.
    */
    ....
}
```
- Changing the associateUser() method to
```PHP
public function associateUser($uID)
{
    $this->setCustomerID($uID);
    $this->save();
}
```
completely fixes this problem, despite being a hack.

# TL;DR
Inconsistent variable between SQL db and object representing it due to what I suppose is bad ORM usage.

## How this Should be Fixed
I am not familiar enough with the ORM or the project to suggest a good fix. That is why this is an issue instead of a pull request. I don't know if changing this would be proper, and I don't know if it would require changing the other methods that use direct SQL queries.

**Comment by Mesuva** [comment]
Thanks for the detailed bug report here, it's much appreciated.

From memory, the associateUser function within Order.php was a bit of a hack/workaround to a problem I faced trying to get it to save the order object correctly with the customer ID at that point in the ordering process.

Your suggestion on how to fix the problem is I think quite valid and not a hack - it's replacing what was a hacky bit of updating in the first place. I actually think the associateUser function is simply not longer needed, and the two lines you've put it in could be simply used directly.

I've made this update on the master with this pull request:
https://github.com/concrete5-community-store/community_store/commit/6ed459cfa59d6ccc52e8b8ac83869aed6745d920

If you find this hasn't solved the issue or you hit further problems please flag this and I'll investigate further.
